### ISSUE TITLE
[v1] Broker_Publish Problem when Signal occurs

### ISSUE THREAD

**Issue by suniltop** [Azure/iot-edge#474]
Hi,
is there any example uart or timer with using signal on linux ?
Broker_Publish doesnt work when signal is occurs from uart

**Comment by aribeironovaes** [comment]
HI @suniltop , 

Unfortunately we don't have an example. We would need more information to check what is going on? 
One thing you could do it to write an example and push a contribution? 

Let us know, 

ANgelo Ribeiro.

**Comment by suniltop** [comment]
[sourcecode.zip](https://github.com/Azure/iot-edge/files/1648863/sourcecode.zip)
Hi angelo,
I prepared a example for this problem
i got a two module with broker that are transmitter and receiver.

Transmitter module has a thread and send a message to receiver per every second.
Message contain is a second value that is increase from timer signal handler (1ms) with prescaler(1000)
When receiver module receive a message, will printing to console

Problem has been occurs randomly and then receiver module dont receive a message







I give an console log below

linaro@tinkerboard:~/remote-debugging$ ./iotbroker iotbroker.json
gateway successfully created from JSON
gateway shall run until ENTER is pressed
Blocking signal 34
Timer ID is 0xb40004d0
Unblocking Signal 34
transmitter Module Timer Configured...
transmitter: Send Message - {Second: 0}
receiver: message received
          message: {Second: 0}
transmitter: Send Message - {Second: 0}
receiver: message received
          message: {Second: 0}
transmitter: Send Message - {Second: 1}
receiver: message received
          message: {Second: 1}
transmitter: Send Message - {Second: 2}
receiver: message received
          message: {Second: 2}
transmitter: Send Message - {Second: 3}
receiver: message received
          message: {Second: 3}
transmitter: Send Message - {Second: 4}
receiver: message received
          message: {Second: 4}
transmitter: Send Message - {Second: 5}
receiver: message received
          message: {Second: 5}
transmitter: Send Message - {Second: 6}
receiver: message received
          message: {Second: 6}
transmitter: Send Message - {Second: 7}
transmitter: Send Message - {Second: 8}
transmitter: Send Message - {Second: 9}
transmitter: Send Message - {Second: 10}
transmitter: Send Message - {Second: 11}

Best Regards

**Comment by damonbarry** [comment]
In your comment, I see that the `receiver: message received` messages stop, but the `transmitter: Send Message ...` messages continue.

Are you printing the transmitter message right next to the `Broker_Publish()` call (either right before or right after)? Or are you printing it earlier in the loop?

I just want to understand whether your transmitter module is actually sending the message vs simply printing that it sent the message when it really didn't. Does `Broker_Publish` return an error?

**Comment by suniltop** [comment]
Hi damonbarry,

Module is writing **transmitter message log** just before calling Broker_Publish

Also i have added transmitter module code that writing return code of broker_publish. 

Broker_Publish returns always 0. This means that there is no error i guess

linaro@tinkerboard:~/test$ ./iotbroker iotbroker.json
gateway successfully created from JSON
gateway shall run until ENTER is pressed
Blocking signal 34
Timer ID is 0xb40004d0
Unblocking Signal 34
transmitter Module Timer Configured...
transmitter: Send Message - {Second: 0}
transmitter: Broker_Publish() Return Code = 0
receiver: message received
          message: {Second: 0}

transmitter: Send Message - {Second: 1}
transmitter: Broker_Publish() Return Code = 0
receiver: message received
          message: {Second: 1}

transmitter: Send Message - {Second: 2}
transmitter: Broker_Publish() Return Code = 0
receiver: message received
          message: {Second: 2}

transmitter: Send Message - {Second: 3}
transmitter: Broker_Publish() Return Code = 0
receiver: message received
          message: {Second: 3}

transmitter: Send Message - {Second: 4}
transmitter: Broker_Publish() Return Code = 0
receiver: message received
          message: {Second: 4}

transmitter: Send Message - {Second: 4}
transmitter: Broker_Publish() Return Code = 0
receiver: message received
          message: {Second: 4}

transmitter: Send Message - {Second: 5}
transmitter: Broker_Publish() Return Code = 0
receiver: message received
          message: {Second: 5}

transmitter: Send Message - {Second: 6}
transmitter: Broker_Publish() Return Code = 0
receiver: message received
          message: {Second: 6}

transmitter: Send Message - {Second: 7}
transmitter: Broker_Publish() Return Code = 0
receiver: message received
          message: {Second: 7}

transmitter: Send Message - {Second: 8}
transmitter: Broker_Publish() Return Code = 0
receiver: message received
          message: {Second: 8}

transmitter: Send Message - {Second: 9}
transmitter: Broker_Publish() Return Code = 0
receiver: message received
          message: {Second: 9}

transmitter: Send Message - {Second: 10}
transmitter: Broker_Publish() Return Code = 0
receiver: message received
          message: {Second: 10}

transmitter: Send Message - {Second: 11}
**transmitter: Broker_Publish() Return Code = 0**

transmitter: Send Message - {Second: 12}
transmitter: Broker_Publish() Return Code = 0

transmitter: Send Message - {Second: 12}
transmitter: Broker_Publish() Return Code = 0

transmitter: Send Message - {Second: 13}
transmitter: Broker_Publish() Return Code = 0

transmitter: Send Message - {Second: 14}
transmitter: Broker_Publish() Return Code = 0

transmitter: Send Message - {Second: 15}
transmitter: Broker_Publish() Return Code = 0

**Comment by suniltop** [comment]
Hi damonbarry,

You are right about my timer is run fast. 
But i have realised that if the signal is generated while the function is running, problem is occurring.
At this point does no matter signal source. ie. uart so,

_If I remove the timer stuff your transmitter and receiver will run happily for as long as I run the gateway_
Yes, you are absolutely right

_Same thing happens if I change the timer to signal 10 times per second_,
No, you are wrong 
Because,
I have made test a lot already different periods (10ms 100ms 1s 10s). There is no solution. :( 
If we chose a big period for timer, we are reduce the likelihood of overlap between these two processes only. For example, I set the timer to 1 second, it occured in 24 hours.

now,
In fact i am not sure that this situation is a problem or not.
i have solved problem several days ago. I have create a thread for every signal

Thanks for your effort

Best Regards

**Comment by damonbarry** [comment]
> i have realised that if the signal is generated while the function is running, problem is occurring.

I think I know what's going on...our broker uses nanomsg, and nanomsg uses sockets to communicate. At least on posix systems, socket reads can be interrupted by other signals that occur on the process. So in this case `nn_recv` returns -1, and `errno==EINTR`. When this happens we should simply call `nn_recv` again, but we don't. Instead we exit the listening thread.

I did a quick patch of the gateway on my Linux box and I've been running your code for about 15 minutes now--it's still going. I'll do some more testing, then I'll see about getting this fix cleaned up and checked in.

**Comment by damonbarry** [comment]
Merged into master (PR #507).
